Local storage v3 migration

grod220 opened the pull request:
Closes https://github.com/prax-wallet/web/issues/166 & quite a bit more. We use chrome.local.storage for persisting some state for the extension. This includes things like wallets, grpc urls, frontend selection, etc. Given local storage is simply a key-value store, it does not come with any tooling to handle migrations. This has made this task a bit more complex than expected.

Our custom migration code for local storage was not quite robust enough to reliably handle the task of the issue. For that reason, I refactored some mechanics. Will add more color inline below. 

Note: This is PR is definitely within the "risky" territory. Migrating user state on the fly can have unintended consequences. This should be throughly tested before we ship another Prax version. For example, how does changing the user's rpc they saved in local storage affect an existing block sync in progress? It's very likely it won't take effect until the service worker is restarted (even though it shows otherwise in the extension). Hence the reason why changing RPCs triggers an extension restart.

grod220 commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,53 @@ export interface StorageItem<T> {
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
```
Some migrations (in our case for V3), requires access to querying the store for other fields. Hence, the migration function is now given a `get` querier.

TalDerei replied:
resolving as outdated

grod220 commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,53 @@ export interface StorageItem<T> {
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
+export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
```
By far the biggest change. Before, migrations had to be:
- From the version of change to the most recent version (v1 -> v3). This means every time we added a new migration, we would have to go to those old versions and make sure they can properly be updated from say (v1 -> v4).  There was not concept of migrating one step at a time.
- The previous map had all migrations for all versions in one object. Was harder to understand (given we tend to think about migrations on a version 43 to version 44 kind of basis).

TalDerei replied:
resolving as outdated

grod220 commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,53 @@ export interface StorageItem<T> {
       return this.defaults[key];
     } else {
-      return await this.migrateIfNeeded(key, result[key]);
+      if (skipMigration) {
+        return result[key].value;
+      } else {
+        return await this.migrateIfNeeded(key, result[key]);
+      }
```
Given some migrations need access to the wider storage getter, we have two ways to get from storage now.

TalDerei replied:
resolving as outdated

grod220 commented on apps/extension/src/storage/migrations/local-v1-migrations.ts:
Copied over from old code.

grod220 commented on apps/extension/src/storage/migrations/local-v2-migrations.ts:
The core migration code that addresses the issue that prompted this whole task.

TalDerei commented:
Isn't it still the case that the recommended rpc entries in the prax registry contain outdated embedded frontends? I think we should remove any frontends that aren't using cron.

grod220 commented on apps/extension/src/storage/base.ts:
```diff
@@ -69,18 +94,47 @@ export class ExtensionStorage<T> {
-    return item.value;
+
+    // Check for an ongoing migration, if exists, return the result of that
+    const ongoingMigration = this.migrationLocks[String(key)] as Promise<T[K]> | undefined;
+    if (ongoingMigration) {
+      return ongoingMigration;
+    }
+
+    // Store the promise in the lock map. Ensures multiple callers don't run migrations multiple times.
+    const migrationPromise = this.migrateField(key, item);
+    this.migrationLocks[String(key)] = migrationPromise;
```
New addition. Given multiple parts of the codebase are accessing fields at the same time that require migration, we need a way to ensure multiple migrations are not run and thus share the same promise to be resolved.

TalDerei replied:
locking mechanism implementation changed, but rationale remains

grod220 commented:
User flows to test:

- Frontend url is in the registry. After ext update: no change made.
- Frontend url is not in the registry. After ext update: changed to a random one in the registry.
- Grpc endpoint is in the registry. After ext update: no change made.
- Grpc endpoint is not in the registry. After ext update: changed to a random one in the registry.
  - Note, block syncing will still pull from previous grpc endpoint until service worker restarts. The UI will still display the old one.
- Go back to April extension version (before v2 migration). V1 -> v2 migration works for both frontend urls and grpc endpoints.

For all of these, confirm no corruption to the other keys (all get bumped to v3 after accessing (when migration takes place)).

TalDerei commented:
The migration is designed for version bumps, but I'm wondering if there's a way to rollback the extension (`V3 --> V2`), which seems to corrupt the local state storage?

<img width="844" alt="Screenshot 2024-08-21 at 9 58 31 AM" src="https://github.com/user-attachments/assets/f609d710-0d8a-4a32-90f9-4193370b9a80">

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
+
+export type VersionSteps = Record<Version, Version>;
 
 export class ExtensionStorage<T> {
```
this is not a suggested code change, but rather a general statement if we decide to restructure our migration mechanism in the future. 

personally, migrations feel more transactional in nature, where migrations for multiple fields should ideally be handled atomically, rather than the current approach of performing migrations on individual fields separately.

one potential approach is a _batch_ migration mechanism, where you define migrations at the version level. It ensures atomicity by checking that all changes associated with a particular version are applied together in a single transaction. so either all migrations succeed or none do

grod220 replied:
This is an important comment. It's unfortunate that chrome.local.storage does not come with any support for versioning or migrations. For this reason, we had to create our own custom migration code. It was probably a bad design that each field has their own version key. A possibly better design is to save a global version number, check for needed migrations when someone attempts a `get()`, and if so, migrate all keys at once.

Going to spend today trying this out. The hardest part of this will be migrating from the old data structure to the new.

TalDerei commented on apps/extension/src/storage/migrations/local-v2-migrations.ts:
```diff
@@ -0,0 +1,78 @@
+import { ChainRegistryClient } from '@penumbra-labs/registry';
+import { sample } from 'lodash';
+
+export const localV2Migrations: MigrationMap<LocalStorageState, LocalStorageState> = {
```
we support both sync / async migration functions – but it's not entirely clear to me how the current implementation explicitly handle errors that might occur during a migration. For instance, if the migration function fails, could this leave the local storage in a partially migrated or inconsistent state? Is there a retry / rollback mechanism?

TalDerei replied:
**update:** : previously, partial state migrations allowed individual fields to have either synchronous or asynchronous migration functions, meaning a migration from one version to another could involve a mix of both sync and async functions. Now that we are migrating the entire storage object for a specified version, it raises a question: do all the fields being migrated have to be either sync or async?

this means that if a single field's migration function is async, the entire migration process needs to be treated as async. This currently doesn't seem to be explicitly enforced in the migration code. 

Additionally, the same error handling comment remains.

grod220 replied:
Talked offline. New storage datastructure does not mix async/sync migrations the way that the former field migrations did. All good here.

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
+export type VersionSteps = Record<Version, Version>;
 
 export class ExtensionStorage<T> {
+  private migrationLocks: Record<string, Promise<T[keyof T]>> = {};
```
thinking more about migration locks, isn't there a risk of blocking if a migration gets stuck or slow? perhaps we should add a timeout mechanism to the migration locks

TalDerei replied:
**update:** the comment should be updated to reflect that the synchronization mechanism prevents concurrent modifications to the _entire_ storage object, not just a single field.

In terms of the risk of blocking, there now might be a risk of blocking between _storage versions_ if the migration gets stuck or slow for a specific storage object (rather than a partial storage key).

grod220 replied:
Talked offline. Not relevant to our specific migration, but think timeouts in the future could be a good feature to add.

TalDerei commented:
tested some workflows – looks good to me

- Grpc endpoint and Frontend url is in the registry. After ext update: no change made.

<img width="857" alt="Screenshot 2024-08-21 at 10 36 02 AM" src="https://github.com/user-attachments/assets/a91ae321-f0c0-416f-a7f7-100ffc79e2fc">

<img width="840" alt="Screenshot 2024-08-21 at 10 40 39 AM" src="https://github.com/user-attachments/assets/059941b9-0947-45d5-a710-dea4b6938187">

----------------------------------------------------------------------------------------------------------------------------

- Frontend url is not in the registry. After ext update: changed to a random one in the registry.

<img width="848" alt="Screenshot 2024-08-21 at 12 19 32 PM" src="https://github.com/user-attachments/assets/ba181a51-2995-4a05-81ab-9bf76b607ef7">

<img width="839" alt="Screenshot 2024-08-21 at 12 21 40 PM" src="https://github.com/user-attachments/assets/d356cc61-5618-44ae-9093-e20736e28f55">

----------------------------------------------------------------------------------------------------------------------------

- Grpc endpoint is not in the registry. After ext update: changed to a random one in the registry

<img width="833" alt="Screenshot 2024-08-21 at 1 21 34 PM" src="https://github.com/user-attachments/assets/11211ab3-3c63-41ab-beb9-157bdea9e27d">

<img width="842" alt="Screenshot 2024-08-21 at 1 25 32 PM" src="https://github.com/user-attachments/assets/ef0a2b74-2da5-4ccb-94b5-a621c5cc542a">

----------------------------------------------------------------------------------------------------------------------------

- If connected a non-mainnet chain id, after ext update: no changes made to grpc endpoint.

<img width="844" alt="Screenshot 2024-08-21 at 1 34 16 PM" src="https://github.com/user-attachments/assets/8d4bd0ad-8c08-4053-b84d-d9bfac7d84ba">

<img width="849" alt="Screenshot 2024-08-21 at 1 36 33 PM" src="https://github.com/user-attachments/assets/59e633fd-e148-4af0-8630-fcc6779064c3">

TalDerei commented on apps/extension/src/storage/migrations/local-v2-migration.test.ts:
```diff
@@ -0,0 +1,143 @@
+
+  describe('grpcEndpoint', () => {
+    test('not set gets ignored', async () => {
+      await v1ExtStorage.set('grpcEndpoint', undefined);
```
what happens in scenarios where the `grpcEndpoint` or `frontendUrl` are _null_ rather than undefined?

grod220 replied:
We don't set those as null, but null fields are possible (though, still of the opinion null is a programming language error)

TalDerei commented:
can you speak on partial storage migrations? suppose we're migrating from `V1` to `V3` through the storage versions, but the process partially fails after completing the migration to `V2`. what generally happens in this case?

grod220 commented:
Will pick up the missing code documentation tomorrow. Just pushed the update+tests that adopts the better global data structure for migrations in local storage. All is working on my end, but want to keep testing user flows given how sensitive of a change this is.

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
```
outdated type in favor of the simplified [MigrationFn](https://github.com/prax-wallet/web/pull/172/commits/8c8922c505bbe919b68e19fdbad026332d442f4f#diff-2bad0a69063227a5d56c8bb08e71d7de75d80fc3bd2e2106ed872311b1bfebc8R18)

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
+    private versionSteps: VersionSteps = {},
   ) {}
 
   async get<K extends keyof T>(key: K): Promise<T[K]> {
```
**update:** `getter` and `setter` kick off the storage migration process for an entire _storage object_ at the version level.

- define a `Version` object with the target version for the storage and map of migrations `RequiredMigrations`, each entry corresponding to separate migration function,
- instantiate the `ExtensionStorage` class,
- call `get` or `set` methods to retrieve or set the values for a collection of keys,
- `get` method internally calls `withDbLock`, checking if the lock is already acquired,
- if the lock is not acquired (ie. `undefined`),  acquire the lock by assigning `dbLock` to the promise returned by `migrateOrInitializeIfNeeded`,
- wait for the lock to resolve, ensuring initialization or migration is complete,
  -  triggers `migrateOrInitializeIfNeeded` and optionally `migrateAllFields` to handle the migration,
- execute the internal get or set operation,
- finally, release the lock.

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -14,75 +15,71 @@ export interface IStorage {
+  // It is quite difficult writing a generic that covers all migration function kinds.
+  // Therefore, the writer of the migration should ensure it is typesafe when they define it.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  0: MigrationFn<any, any>;
```
should we make this an _optional_ field? for instance, migration function for version 0 won't be relevant for a higher version.

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -14,75 +15,71 @@ export interface IStorage {
 
-export class ExtensionStorage<T> {
-  private migrationLocks: Record<string, Promise<T[keyof T]>> = {};
+export class ExtensionStorage<T extends { dbVersion: number }> {
```
can we include a description of the properties for the ExtensionStorage class:
- responsible for managing the storage system – storage versioning and migrations
- the state managed by this class explicitly includes versioning information from `dbVersion`

TalDerei commented on apps/extension/src/storage/migrations/local-v1-migration.ts:
```diff
@@ -0,0 +1,132 @@
+}
+
+// Update LocalStorageState to V1LocalStorageState if there is a version bump
+export const localV0Migration: MigrationFn<V0LocalStorageState, LocalStorageState> = v0 => {
```
`localV0Migration` handles the migration from the old `V0LocalStorageState` structure, where each field is wrapped in a `StorageItem`, to the new `LocalStorageState` structure. However, we don't have `localV1Migration: MigrationFn<LocalStorageState, LocalStorageState>` to test migrations within the `LocalStorageState` structure itself.

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -69,18 +92,29 @@ export class ExtensionStorage<T> {
   }
 
   private async migrateIfNeeded<K extends keyof T>(key: K, item: StorageItem<T[K]>): Promise<T[K]> {
```
resolving as outdated

TalDerei commented on apps/extension/src/storage/base.ts:
```diff
@@ -69,18 +92,29 @@ export class ExtensionStorage<T> {
-        await this.set(key, transformedVal);
-        return transformedVal;
-      } else {
-        // If there's no migration function, handle it appropriately, possibly by just returning the current value
-        return item.value;
```
resolving as outdated

TalDerei commented:
> tested some workflows – looks good to me

I still haven't retested these workflows post refactor

grod220 commented:
Tested on all user flows, things in my view are looking good ✅

TalDerei reviewed (approved):
left a minor comment; double-checked the locking system logic and that we're propagating the errors up the caller stack properly, and everything looks good to me.

TalDerei commented:
tested user workflows – looks good to me ✔️

* Grpc endpoint and Frontend url is in the registry. After ext update: no change made.

<img width="848" alt="Screenshot 2024-08-24 at 3 16 54 PM" src="https://github.com/user-attachments/assets/9895f69c-df4e-4207-b8ef-7c176fc3f165">

<img width="959" alt="Screenshot 2024-08-24 at 3 19 15 PM" src="https://github.com/user-attachments/assets/dd0fef34-ae91-4967-a505-8eac9c658319">

----------------------------------------------------------------------------------------------------------------------------

* Frontend url is not in the registry. After ext update: changed to a random one in the registry.

<img width="838" alt="Screenshot 2024-08-24 at 3 32 33 PM" src="https://github.com/user-attachments/assets/29198e67-0a35-47e9-8709-b0a40ec9cd96">

<img width="963" alt="Screenshot 2024-08-24 at 3 35 32 PM" src="https://github.com/user-attachments/assets/1b31385c-4ad7-4ac5-89f7-ba6923a0f393">

----------------------------------------------------------------------------------------------------------------------------

* Grpc endpoint is not in the registry. After ext update: changed to a random one in the registry

<img width="841" alt="Screenshot 2024-08-24 at 6 27 43 PM" src="https://github.com/user-attachments/assets/87a08521-e3ed-499b-94af-cc26d5b76821">

<img width="969" alt="Screenshot 2024-08-24 at 6 29 40 PM" src="https://github.com/user-attachments/assets/0bcd5744-dc91-42ca-82ba-7c43035582f6">

----------------------------------------------------------------------------------------------------------------------------
 
* If connected a non-mainnet chain id, after ext update: no changes made to grpc endpoint.

<img width="854" alt="Screenshot 2024-08-24 at 6 50 53 PM" src="https://github.com/user-attachments/assets/345241b7-d5d9-4755-87c7-b26bb6f793df">

<img width="962" alt="Screenshot 2024-08-24 at 6 53 38 PM" src="https://github.com/user-attachments/assets/c210045d-0b81-477b-a935-55e08fdfbb75">
